src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/CosmasInMemoryService.kt

```
@@ -76,11 +84,65 @@ class CosmasInMemoryService : CosmasGrpc.CosmasImplBase() {
+                        "You ask for version ${request.version} that is negative")
+        }
+        return status
+    }
+
     private fun addNewVersion(request: CosmasProto.CreateVersionRequest) {
         synchronized(this.files) {
             val fileVersions = this.files[request.fileId] ?: mutableListOf()
             fileVersions.add(request.file)
             this.files[request.fileId] = fileVersions
         }
     }
+
+    override fun createHistoryOfVersion(request: CosmasProto.CreateHistoryOfVersionRequest,
+                                        responseObserver: StreamObserver<CosmasProto.CreateHistoryOfVersionResponse>) {
+        println("Get request for create new history of file # ${request.fileId}")
+        synchronized(this.versions) {
+            val historyOfFile = versions[request.fileId] ?: mutableListOf()
+            historyOfFile.add(storedStructure(request.user, request.patch, request.timeStamp))
+            versions[request.fileId] = historyOfFile
+        }
+        val response: CosmasProto.CreateHistoryOfVersionResponse = CosmasProto.CreateHistoryOfVersionResponse
+                .newBuilder()
+                .build()
+        responseObserver.onNext(response)
+        responseObserver.onCompleted()
+    }
+
+    override fun getHistoryOfVersion(request: CosmasProto.GetHistoryOfVersionRequest,
+                                      responseObserver: StreamObserver<CosmasProto.GetHistoryOfVersionResponse>) {
+        println("Get request for history ${request.version} file # ${request.fileId}")
+        val response = CosmasProto.GetHistoryOfVersionResponse.newBuilder()
+        synchronized(this.versions) {
+            val fileHistoryOfVersions = this.versions[request.fileId]
+            val requestStatus = verifyGetHistoryOfVersionRequest(request)
+            if (requestStatus.isOk) {
+                val history = fileHistoryOfVersions?.get(request.version)
+                response.user = history?.user

```

If you received ok, you know for sure that history is not null. If it suddenly becomes null, it is an error which must be reported. Report it, do not silently return success with empty structure to the client.